frontend/src/data/utils/DatasetCalculator.ts

```diff
@@ -0,0 +1,128 @@
+import { DataFrame, IDataFrame } from "data-forge";
+import { BreakdownVar } from "../query/Breakdowns";
+import { ALL, UNKNOWN, UNKNOWN_RACE } from "./Constants";
+import { applyToGroups } from "./datasetutils";
+
+export class DatasetCalculator {
+  /** Calculates a rate as occurrences per 100k */
+  per100k(numerator: number, denominator: number): number | null {
+    return numerator == null || denominator == null || denominator === 0
+      ? null
+      : Math.round(100000 * (numerator / denominator));
+  }
+
+  /** Calculates a rate as a percent to one decimal place. */
+  percent(numerator: number, denominator: number): number | null {
+    return numerator == null || denominator == null || denominator === 0
+      ? null
+      : Math.round((1000 * numerator) / denominator) / 10;
+  }
+
+  /** Finds expected value of an ailment based on a population sample. */
+  estimateTotal(
+    sample_percentage: number,
+    total_population: number
+  ): number | null {
+    return sample_percentage == null ||
+      total_population == null ||
+      total_population === 0
+      ? null
+      : Math.round((sample_percentage / 100) * total_population);
+  }
+
+  /**
+   * Calculates a percent share column. In order for this to work, a "Total"
+   * value must be present for each group being applied to.
+   * Note that this function is not very efficient so it should only be used on
+   * small batches of data.
+   * @param df The DataFrame to apply to
+   * @param rawCountCol The name of the column with the raw count
+   * @param pctShareCol The name of the column to create with percent share
+   * @param breakdownCol The name of the column to calculate the percent across.
+   * @param groupByCols The columns to group by before calculating the total.
+   */
+  calculatePctShare(
+    df: IDataFrame,
+    rawCountCol: string,
+    pctShareCol: string,
+    breakdownCol: BreakdownVar,
+    groupByCols: string[]
+  ) {
+    return applyToGroups(df, groupByCols, (group) => {
+      const totalRow = group.where((r) => r[breakdownCol] === ALL);
+      if (totalRow.count() === 0) {
+        throw new Error("No Total value for group");
+      }
+      const total = totalRow.first()[rawCountCol];
+      return group
+        .generateSeries({
+          [pctShareCol]: (row) => this.percent(row[rawCountCol], total),
+        })
+        .resetIndex();
+    });
+  }
+
+  calculatePctShareOfKnown(
```

Comment:
It would be good to add a test for this function as you are adding tests for all the others

Reply:
okie